transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java

```diff
@@ -160,9 +162,112 @@ NativeDatagramPacketArray cleanDatagramPacketArray() {
         return datagramPacketArray;
     }
 
+    @Override
+    protected void executeScheduledRunnable(ScheduledTaskRunnable runnable) {
+        if (runnable.isAddition()) {
+            try {
+                trySetTimerFd(runnable);
+            } catch (IOException cause) {
+                throw new RejectedExecutionException(cause);
+            }
+        }
+        // else this is a removal of scheduled task and we could attempt to detect if this task was responsible
+        // for the next delay, and find the next lowest delay in the queue to re-set the timer. However this
+        // is not practical for the following reasons:
+        // 1. The data structure is a PriorityQueue, and the scheduled task has not yet been removed. This means
+        //    we would have to add/remove the head element to find the "next timeout".
+        // 2. We are not on the EventLoop thread, and the PriorityQueue is not thread safe. We could attempt
+        //    to do (1) if we are on the EventLoop but when the EventLoop wakes up it checks if the timeout changes
+        //    when it is woken up and before it calls epoll_wait again and adjusts the timer accordingly.
+        // The result is we wait until we are in the EventLoop and doing the actual removal, and also processing
+        // regular polling in the EventLoop too.
+
+        execute(runnable);
+    }
+
+    @Override
+    protected boolean wakesUpForTask(Runnable task) {
+        return super.wakesUpForTask(task) || !(task instanceof ScheduledTaskRunnable);
```

Comment:
@Scottmitch maybe I am missing something but is this really correct ? I thought it should be:

```
return !(task instanceof ScheduledTaskRunnable) && super.wakesUpForTask(task);
```

At least from what I understand we should never need to wakeup for a `ScheduledTaskRunnable` as the timerfd will take care of it.

Reply:
Yes you are correct ... the logic here is flawed 🤦‍♂️